api/runtime/sandbox/v1/sandbox.proto

```diff
@@ -36,6 +37,10 @@ service Sandbox {
 	// StartSandbox will start previsouly created sandbox.
 	rpc StartSandbox(StartSandboxRequest) returns (StartSandboxResponse);
 
+	// Platform queries the platform the sandbox is going to run containers on.
+	// containerd will use this to generate a proper OCI spec.
+	rpc Platform(PlatformRequest) returns (PlatformResponse);
```

Comment:
Let's say LCOW implements Platform, does it return Linux?

Reply:
I wonder if we can add this as part of the configuration (perhaps in addition?). Greedily, it makes me think that this needn't be sandbox API only.

Reply:
Yea that was my first thought here https://github.com/containerd/containerd/issues/7910, if it's generic though and specified here we still need to forward it to the shim somehow (annotation would be easiest but ideally whatever would denote it would be standardized/part of the runtime spec)

Reply:
It could to a degree, would just be kind of a hack. So on Windows to denote LCOW we'd fill in the Windows portion of the spec so it was non-nil and then fill in the Linux portion as well. Seems like to avoid needing to do some weird matrix of that (e.g check if host OS part of spec is filled in and also guest) we could just have a documented approach to follow, like check the "io.containerd.targetos" annotation or something. Also for the LCOW approach I'm not sure what you'd do for guests that don't have a resulting field in the runtime spec like Darwin or FreeBSD.

Reply:
Hmm. I didn't realize that LCOW and WCOW were in the same shim. Would splitting into two shims even an option?

If not, we may need to discuss the modeling first. Is Sandbox:Platform = 1:1? Can we make it by starting a sandbox/shim with a platform? Is it too limiting from hcsshim's standpoint?

Reply:
@kzys hcsshim currently does not implement sandbox API, just Task Service. So it's tricky to support multiple platforms (and this is what I want to address with sandbox API). The way this intended to work is something like:

```

- CRI.CreatePodSandbox(id=1) - call SandboxAPI.CreateSandbox (targetOS=Windows)
- CRI.CreatePodSandbox(id=2) - call SandboxAPI.CreateSandbox (targetOS=Linux)

// Implemented by this PR.
- CRI.CreateContainer(sandboxId=1) - SandboxAPI.Platform returns Windows - we build Windows spec
- CRI.CreateContainer(sandboxId=2) - SandboxAPI.Platform returns Linux - will build Linux spec

```

Reply:
Note here is that we still need to figure out how to pass targetOS to hcsshim (https://github.com/containerd/containerd/issues/7910)

Reply:
In kubernetes, currently, we can only determine the runtime platform by RuntimeHandler and annotation.

If the hcsshim supports multiple platforms, maybe we can add different CRI RuntimeHandlers with specific annotation so that the shim will know that it is going to ship containers on which kind of the platform.

IMO, it is fine to pass the `targetOS` as field in the `CreateSandbox` in controller, because the sandbox controller is extra layer to handle sandbox. But in the runtime layer, I think we should keep `Sandbox:Platform=1:1` concept as it is easy to understand and maintained.

Reply:
> Hmm. I didn't realize that LCOW and WCOW were in the same shim. Would splitting into two shims even an option?

It may be, but not sure that's the right move to plop the burden on shim authors. If we can somehow pass what targetOS to expect that should remedy this a ton, but I don't think having to have N separate shims for N supported guests would be sane for authors

Reply:
@fuweid I agree completely, although I'd prefer if whatever field we devise is not windows/hcsshim/etc. specific and could be used for all shim authors. The model in my head is basically what you wrote, for CRI we can have multiple runtime handlers (which is what Windows does anyways) that all point to the same underlying shim, but they can have different options defined that state what target platform this runtime supports. We could parse that out and pass it in CreateSandbox for shims that implement the Sandbox API. For shims that don't implement the Sandbox API perhaps we should set an annotation for them. 

So in your pod spec the divide is just mycoolruntime-linux vs. mycoolruntime-windows as your runtime handler and you are off to the races. This doesn't solve how to unpack images for the target platform on your host, but that's way outside the scope of this change.

Reply:
Anyways I think Maksym's change is squarely focused on spitting out the right spec for individual containers, so maybe we're getting into the weeds. I'll circle back to this thread after reviewing here as maybe it's impossible to separate the two problems hahah

Reply:
Sorry for not getting a chance to comment on this sooner. Adding a few thoughts from my side.

LCOW isn't supported on CRI today, we don't really need to solve for that at this point. However, we should aim for a path where the runhcs shim can implement sandbox support, and right now I don't think it would have a good way to know what platform to report.

Given that we will probably need some way (e.g. a platform field on each runtime handler like in #7910) for CRI to tell the shim what platform to create a sandbox for, it seems a bit odd to me that we then have to turn around and ask the sandbox what its platform is. We just got done telling it what platform it should be, so can we just remember that value (perhaps stored in CRI's sandboxstore) rather than querying it?

This would mean we can avoid adding a new controller/sandbox API to query what platform is running, which simplifies the changes needed here.

> Would splitting into two shims even an option?

I'd prefer we don't have to force a change in shim architecture to accomodate this. There are also some technical reasons to want a single shim binary, such as code page sharing between shim processes.

Reply:
> it seems a bit odd to me that we then have to turn around and ask the sandbox what its platform is. 

That's fair. However there is no official way on CRI level to specify target platform, so it's currently up to a runtime how to figure out which target OS to run (configs, annotations, etc). So I'd prefer to avoid hard-coding this on CRI side for now. This is a good path forward though, I like this, we can update interfaces in 2.0 as sandbox API is still experimental and we can tune interface as we need.

Reply:
Well I guess this is okay 😆 Personally feel that what platform is being run is a central enough part that CRI should explicitly  specify it, but don't think it's a big deal.

Reply:
KEP time? 🤓

Reply:
Well, it would be a configuration option - not hard coded, similar to runtime's config path or which CNI plugin to use, right?